File: examples/rq/monitor.py
"""
Queue Monitor - RQ Integration Example

This example shows how to integrate Queue Monitor with Python RQ
to get alerted when queues back up, jobs fail, or workers go down.

Installation:
    pip install rq redis

Usage:
    # In a separate terminal, start a worker:
    python worker.py

    # Then enqueue jobs:
    python enqueue.py

    # Queue Monitor will alert you when:
    # - Queue depth exceeds threshold
    # - Jobs fail repeatedly
    # - Workers stop heartbeating
"""

import os
import redis
from rq import Queue
from rq.worker import Worker
import requests
import json
from datetime import datetime, timedelta
from typing import Optional


class QueueMonitor:
    """
    Monitor RQ queues and send alerts when thresholds are exceeded.

    This is a reference implementation. Queue Monitor will provide
    a hosted service with Slack, email, and webhook notifications.
    """

    def __init__(
        self,
        redis_url: str = "redis://localhost:6379",
        alert_webhook: Optional[str] = None,
        queue_depth_threshold: int = 1000,
        check_interval: int = 60,
    ):
        self.redis = redis.from_url(redis_url)
        self.alert_webhook = alert_webhook or os.getenv("QUEUE_MONITOR_WEBHOOK")
        self.queue_depth_threshold = queue_depth_threshold
        self.check_interval = check_interval
        self.last_alert = {}  # Track last alert time per queue

    def check_queue_depth(self, queue_name: str) -> dict:
        """Check if queue depth exceeds threshold."""
        key = f"rq:queue:{queue_name}"
        length = self.redis.llen(key)

        status = {
            "queue": queue_name,
            "depth": length,
            "threshold": self.queue_depth_threshold,
            "status": "ok" if length < self.queue_depth_threshold else "alert",
            "timestamp": datetime.utcnow().isoformat(),
        }

        if status["status"] == "alert":
            # Check if we already alerted recently (avoid spam)
            last_alert_time = self.last_alert.get(f"{queue_name}:depth")
            if not last_alert_time or datetime.utcnow() - last_alert_time > timedelta(minutes=15):
                self.send_alert(status)
                self.last_alert[f"{queue_name}:depth"] = datetime.utcnow()

        return status

    def check_failed_jobs(self, queue_name: str) -> dict:
        """Check for failed jobs."""
        failed_key = f"rq:queue:{queue_name}:failed"
        failed_count = self.redis.scard(failed_key)

        status = {
            "queue": queue_name,
            "failed_count": failed_count,
            "status": "ok" if failed_count == 0 else "warning",
            "timestamp": datetime.utcnow().isoformat(),
        }

        if failed_count > 0:
            last_alert_time = self.last_alert.get(f"{queue_name}:failed")
            if not last_alert_time or datetime.utcnow() - last_alert_time > timedelta(minutes=30):
                self.send_alert(status)
                self.last_alert[f"{queue_name}:failed"] = datetime.utcnow()

        return status

    def check_workers(self, queue_name: str) -> dict:
        """Check if workers are alive."""
        workers = Worker.all(queue=Queue(queue_name, connection=self.redis), connection=self.redis)
        alive_workers = [w for w in workers if w.state != "stopped"]

        status = {
            "queue": queue_name,
            "workers": len(alive_workers),
            "status": "ok" if len(alive_workers) > 0 else "critical",
            "timestamp": datetime.utcnow().isoformat(),
        }

        if status["status"] == "critical":
            last_alert_time = self.last_alert.get(f"{queue_name}:workers")
            if not last_alert_time or datetime.utcnow() - last_alert_time > timedelta(minutes=5):
                self.send_alert(status)
                self.last_alert[f"{queue_name}:workers"] = datetime.utcnow()

        return status

    def send_alert(self, status: dict) -> None:
        """Send alert to webhook (Slack, Discord, etc)."""
        message = self.format_alert(status)
        print(f"[ALERT] {message}")

        if self.alert_webhook:
            try:
                requests.post(
                    self.alert_webhook,
                    json={"text": message},
                    timeout=5,
                )
            except requests.RequestException as e:
                print(f"[ERROR] Failed to send alert: {e}")

    def format_alert(self, status: dict) -> str:
        """Format alert message."""
        if "depth" in status:
            return f"🚨 Queue '{status['queue']}' depth: {status['depth']} (threshold: {status['threshold']})"
        elif "failed_count" in status:
            return f"⚠️ Queue '{status['queue']}' has {status['failed_count']} failed jobs"
        elif "workers" in status:
            return f"🔴 Queue '{status['queue']}' has no active workers!"
        return str(status)

    def monitor(self, queue_names: list[str]) -> None:
        """Continuously monitor queues."""
        print(f"Monitoring queues: {', '.join(queue_names)}")
        print(f"Depth threshold: {self.queue_depth_threshold}")
        print(f"Check interval: {self.check_interval}s")
        print("-" * 50)

        while True:
            import time

            for queue_name in queue_names:
                self.check_queue_depth(queue_name)
                self.check_failed_jobs(queue_name)
                self.check_workers(queue_name)

            time.sleep(self.check_interval)


# Example usage
if __name__ == "__main__":
    import sys

    # Get queue names from command line or use default
    queues = sys.argv[1:] if len(sys.argv) > 1 else ["default", "high", "low"]

    monitor = QueueMonitor(
        redis_url=os.getenv("REDIS_URL", "redis://localhost:6379"),
        queue_depth_threshold=500,
        check_interval=60,
    )

    try:
        monitor.monitor(queues)
    except KeyboardInterrupt:
        print("\nMonitoring stopped.")

File: examples/bullmq/monitor.ts
/*
 * Queue Monitor - BullMQ Integration Example
 *
 * This example shows how to integrate Queue Monitor with BullMQ
 * to get alerted when queues back up, jobs fail, or workers go down.
 *
 * Installation:
 *   npm install bullmq ioredis
 *
 * Usage:
 *   # Run monitor
 *   ts-node monitor.ts
 *
 *   # Queue Monitor will alert you when:
 *   # - Queue depth exceeds threshold
 *   # - Jobs fail repeatedly
 *   # - Workers stop processing
 */

import { Queue, Worker, Job } from 'bullmq';
import Redis from 'ioredis';

interface AlertStatus {
  queue: string;
  status: 'ok' | 'warning' | 'critical';
  timestamp: string;
  [key: string]: any;
}

interface MonitorConfig {
  redisUrl?: string;
  alertWebhook?: string;
  queueDepthThreshold?: number;
  checkInterval?: number;
}

class BullMQMonitor {
  private connection: Redis;
  private alertWebhook: string | undefined;
  private queueDepthThreshold: number;
  private checkInterval: number;
  private lastAlerts: Map<string, Date> = new Map();
  private queues: Map<string, Queue> = new Map();

  constructor(config: MonitorConfig = {}) {
    this.connection = new Redis(config.redisUrl || 'localhost:6379');
    this.alertWebhook = config.alertWebhook || process.env.QUEUE_MONITOR_WEBHOOK;
    this.queueDepthThreshold = config.queueDepthThreshold || 1000;
    this.checkInterval = config.checkInterval || 60000;
  }

  async getQueue(queueName: string): Promise<Queue> {
    if (!this.queues.has(queueName)) {
      this.queues.set(queueName, new Queue(queueName, { connection: this.connection }));
    }
    return this.queues.get(queueName)!;
  }

  async checkQueueDepth(queueName: string): Promise<AlertStatus> {
    const queue = await this.getQueue(queueName);
    const waiting = await queue.getWaitingCount();
    const active = await queue.getActiveCount();
    const delayed = await queue.getDelayedCount();
    const failed = await queue.getFailedCount();

    const totalDepth = waiting + active + delayed;

    const status: AlertStatus = {
      queue: queueName,
      depth: totalDepth,
      waiting,
      active,
      delayed,
      failed,
      threshold: this.queueDepthThreshold,
      status: totalDepth > this.queueDepthThreshold ? 'alert' : 'ok',
      timestamp: new Date().toISOString(),
    };

    if (status.status === 'alert') {
      const lastAlert = this.lastAlerts.get(`${queueName}:depth`);
      if (!lastAlert || Date.now() - lastAlert.getTime() > 15 * 60 * 1000) {
        await this.sendAlert(status);
        this.lastAlerts.set(`${queueName}:depth`, new Date());
      }
    }

    return status;
  }

  async checkFailedJobs(queueName: string): Promise<AlertStatus> {
    const queue = await this.getQueue(queueName);
    const failed = await queue.getFailedCount();

    const status: AlertStatus = {
      queue: queueName,
      failedCount: failed,
      status: failed > 0 ? 'warning' : 'ok',
      timestamp: new Date().toISOString(),
    };

    if (failed > 0) {
      const lastAlert = this.lastAlerts.get(`${queueName}:failed`);
      if (!lastAlert || Date.now() - lastAlert.getTime() > 30 * 60 * 1000) {
        await this.sendAlert(status);
        this.lastAlerts.set(`${queueName}:failed`, new Date());
      }
    }

    return status;
  }

  async checkWorkers(queueName: string): Promise<AlertStatus> {
    const queue = await this.getQueue(queueName);
    const workers = await queue.getWorkers();

    const status: AlertStatus = {
      queue: queueName,
      workers: workers.length,
      status: workers.length > 0 ? 'ok' : 'critical',
      timestamp: new Date().toISOString(),
    };

    if (status.status === 'critical') {
      const lastAlert = this.lastAlerts.get(`${queueName}:workers`);
      if (!lastAlert || Date.now() - lastAlert.getTime() > 5 * 60 * 1000) {
        await this.sendAlert(status);
        this.lastAlerts.set(`${queueName}:workers`, new Date());
      }
    }

    return status;
  }

  async sendAlert(status: AlertStatus): Promise<void> {
    const message = this.formatAlert(status);
    console.log(`[ALERT] ${message}`);

    if (this.alertWebhook) {
      try {
        await fetch(this.alertWebhook, {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify({ text: message }),
        });
      } catch (error) {
        console.error(`[ERROR] Failed to send alert:`, error);
      }
    }
  }

  formatAlert(status: AlertStatus): string {
    if ('depth' in status) {
      return `🚨 Queue '${status.queue}' depth: ${status.depth} (threshold: ${status.threshold})`;
    }
    if ('failedCount' in status) {
      return `⚠️ Queue '${status.queue}' has ${status.failedCount} failed jobs`;
    }
    if ('workers' in status) {
      return `🔴 Queue '${status.queue}' has no active workers!`;
    }
    return JSON.stringify(status);
  }

  async monitor(queueNames: string[]): Promise<never> {
    console.log(`Monitoring BullMQ queues: ${queueNames.join(', ')}`);
    console.log(`Depth threshold: ${this.queueDepthThreshold}`);
    console.log(`Check interval: ${this.checkInterval}ms`);
    console.log('-'.repeat(50));

    while (true) {
      for (const queueName of queueNames) {
        await this.checkQueueDepth(queueName);
        await this.checkFailedJobs(queueName);
        await this.checkWorkers(queueName);
      }

      await new Promise(resolve => setTimeout(resolve, this.checkInterval));
    }
  }

  async close(): Promise<void> {
    for (const queue of this.queues.values()) {
      await queue.close();
    }
    this.connection.quit();
  }
}

// Example usage
async function main() {
  const queueNames = process.argv.slice(2);
  const queuesToMonitor = queueNames.length > 0 ? queueNames : ['default', 'high', 'low'];

  const monitor = new BullMQMonitor({
    redisUrl: process.env.REDIS_URL || 'localhost:6379',
    queueDepthThreshold: 500,
    checkInterval: 60000,
  });

  try {
    await monitor.monitor(queuesToMonitor);
  } catch (error) {
    console.error('Monitor error:', error);
    await monitor.close();
    process.exit(1);
  }
}

if (require.main === module) {
  main().catch(console.error);
}

export { BullMQMonitor, AlertStatus, MonitorConfig };
